libs/common/NumUtils.cs

```diff
@@ -115,21 +116,18 @@ public static bool TryBytesToLong(int length, byte* source, out long result)
         {
             var fNeg = *source == '-';
             var beg = fNeg ? source + 1 : source;
-            var end = source + length;
-            var digit = *beg - '0';
+            var len = fNeg ? length - 1 : length;
             result = 0;
 
-            // Check first digit which needs to be non-zero
-            if (digit is <= 0 or > 9)
+            // Do not allow leading zeros
+            if (len > 1 && *beg == '0')
                 return false;
 
-            while (beg < end)
-            {
-                digit = *beg++ - '0';
-                if (digit is < 0 or > 9)
-                    return false;
-                checked { result = (result * 10) + digit; }
-            }
+            // Parse number and check consumed bytes to avoid alphanumeric strings
+            if (!Utf8Parser.TryParse(new Span<byte>(beg, len), out result, out var bytesConsumed) || bytesConsumed != len)
```

Comment:
Won't `Utf8Parser.TryParse(ReadOnlySpan<Byte>, Int64, Int32, Char)` take care of the leading sign? If we allow leading zeros we can get rid of the previous checks as well and just fully map to `Utf8Parser.TryParse()`.

Reply:
Is `Utf8Parser.TryParse` better or more efficient than `TryBytesToLong`? Historically, these APIs have way too many checks that might not be relevant to our needs.

Reply:
I benchmarked this earlier, and we can do better than `Utf8Parser.TryParse` by relaxing the requirements a bit. Maybe keep the existing code and make `digit `explicitly unsigned to simplify the bounds check (`digit <= 9`).

Reply:
> Won't `Utf8Parser.TryParse(ReadOnlySpan<Byte>, Int64, Int32, Char)` take care of the leading sign? If we allow leading zeros we can get rid of the previous checks as well and just fully map to `Utf8Parser.TryParse()`.

In both cases, the problem is the leading zeros. 
If the input is something like -010, bytesConsumed will be 4 so we cannot identify that this is not a valid number string. Same applies if it not a negative number (e.g. 010 will return 3 bytesConsumed).
So I broke it down into two distinct checks.
Redis does not allow leading zeros.

![image](https://github.com/microsoft/garnet/assets/96085550/4bc73953-d556-46e4-ab40-d3be735e7cc2)

Performance wise it does not make much of a difference. check PR #262 for perf comparison.
Generally, I noticed there is +/-50ns difference.
Also checked the TryParseCustom vs Utf8Parser and it seems the former is better for numbers with few digits and the latter is better for numbers with many digits.
I think that having a custom parser could be better for reading headers of commands since there we expect few digits most of the time.
Here it should not make much difference, because it does not also touch the critical path.